### PR TITLE
CI: Validate config export

### PR DESCRIPTION
**Pull request by zeshanziya**
Validate config using `ddev drush config:status` and add a PR comment if there are differences.

### CODE REVIEW

**Comment by zeshanziya** [comment]
> @zeshanziya, this is a good start. I notice this doesn't handle the situation you described where the config import reverts the changes by update-db.

@hussainweb I didn't really get a clear way to proceed. So, I am thinking instead of deploying using `drush deploy`, which runs `updatedb` first and then `config:import`, we can first run `config:import` and then `updatedb`. I'm not sure about any issues caused by running `config:import` before `updatedb`. I will check and revert.

**Comment by hussainweb** [comment]
@zeshanziya the link I shared in slack gives a specific sequence. How about using that?

**Comment by zeshanziya** [comment]
@hussainweb I checked it and found it similar to `drush deploy` steps. its just that it runs `config:import` twice.

**Comment by hussainweb** [comment]
Yes, and it runs update db once more after config import. This solves the problem you had raised earlier.

**Comment by zeshanziya** [comment]
That link is deprecated in reference to the new ADR.

Old ADR: https://architecture.lullabot.com/adr/20210924-drupal-build-steps/
New ADR: https://architecture.lullabot.com/adr/20230929-drupal-build-steps/

Even if we follow the old ADR steps, running updatedb again will not run those updates again. I see that on the first run, it does not execute post_update hooks, which are executed on the second run. However, we may have updates using `hook_update_N` which are executed on the first run only.

I have modified the CI step. Can you please have a look?

**Comment by zeshanziya** [comment]
@hussainweb I have finalized the changes. Please review.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -120,17 +123,68 @@ jobs:
         run: |
           ddev composer install
           ddev pull platform -y
-          ddev drush deploy -y
+
+      - name: Run phpstan
+        run: ddev php ./vendor/bin/grumphp run --tasks=phpstan --no-interaction
```
Comment (hussainweb) [inline .github/workflows/ci.yml:128]: @zeshanziya, I see you have made this to the top, and that makes sense, but let's do that in another PR? I want this PR to be an example of change we can use elsewhere as well.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -120,17 +123,68 @@ jobs:
...
+
+      - name: Rebuild cache
+        run: |
+          ddev drush cache:rebuild -y
+          ddev drush config:status
```
Comment (hussainweb) [inline .github/workflows/ci.yml:133]: Was this for debugging? Is it useful to keep these outputs in the logs?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -120,17 +123,68 @@ jobs:
...
+
+      - name: Import Config and Run updates
+        run: |
+          ddev drush updatedb --no-cache-clear --yes
+          ddev drush config:import --yes
+          ddev drush cache:rebuild --yes
+          ddev drush deploy:hook --yes
```
Comment (hussainweb) [inline .github/workflows/ci.yml:140]: This is exactly the order documented in `drush deploy`. What's the value of breaking it down if we are going to use the same sequence?

---

From https://www.drush.org/12.x/deploycommand/

```
drush updatedb --no-cache-clear
drush cache:rebuild
drush config:import
drush cache:rebuild
drush deploy:hook
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -120,17 +123,68 @@ jobs:
...
+
+      # Validate config changes
+      - name: Validate Config Export
+        id: validate_config
+        run: |
+          status_output=$(ddev drush config:status 2>&1)
+          echo "$status_output"
+          echo "status_output<<EOF" >> $GITHUB_OUTPUT
+          echo "$status_output" >> $GITHUB_OUTPUT
+          echo "EOF" >> $GITHUB_OUTPUT
```
Comment (hussainweb) [inline .github/workflows/ci.yml:150]: Some comments might be helpful here. We will come back looking at this change in the future, I'm sure.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -120,17 +123,68 @@ jobs:
...
+            ```
+
+      - name: Fail if Config Export is not clean
+        if: ${{ !contains(steps.validate_config.outputs.status_output, 'No differences') }}
+        run: |
+          echo "Config export has differences"
+          exit 1
+
+      - name: Deploy Site
+        run: ddev drush deploy -y
```
Comment (hussainweb) [inline .github/workflows/ci.yml:180]: We have already run this once before. Why again?

File: .github/workflows/ci.yml (hunk 3 above)
Comment (zeshanziya) [reply .github/workflows/ci.yml:140]: We don't want drush cache:rebuild to run after updatedb. After updatedb, any config changes in update hooks are written to the database, but the cache is not cleared. As a result, config:import will only import any exported config that has been pushed and is different. If we clear the cache, the config cache is updated. If the config export is not pushed, it will show a difference in the config, and the old config will be imported.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (zeshanziya) [reply .github/workflows/ci.yml:133]: This is to rebuild the cache and config cache. It seems cache:rebuild is not rebuilding the config cache. After running config:status, it recaches all configs. We are pulling only the database from Platform.sh, and Redis is empty on DDEV start. So, I am rebuilding all the caches after the pull.

**Comment by zeshanziya** [comment]
@hussainweb I have made changes with new logic. Please review. I am seeing if I can optimize that Validate config step. should I put it in a separate script file? 

I am doing some testing in a separate PR #545

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -118,6 +121,74 @@ jobs:
...
+          # run updates and build cache
+          ddev drush updatedb --yes
+          ddev drush cache:rebuild --yes
+
+          # identify set of configuration changes after running updatedb and store it in a text file.
+          ddev drush config:status --fields name --format list | awk NF > ${{ runner.temp }}/updated_config.txt
+
+          # sort the lines within files as comm requires sorted files
+          sort ${{ runner.temp }}/initial_config.txt -o ${{ runner.temp }}/sorted_initial_config.txt
+          sort ${{ runner.temp }}/updated_config.txt -o ${{ runner.temp }}/sorted_updated_config.txt
```
Comment (hussainweb) [inline .github/workflows/ci.yml:144]: Instead of a separate `sort`, let's pipe `awk` output into `sort`.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -118,6 +121,74 @@ jobs:
...
+      # Update the comment only if it was added at first place
+      - name: Update the comment on success
+        if: ${{ contains(steps.validate_config.outputs.status_output, 'No differences') }}
+        uses: mshick/add-pr-comment@v2
+        with:
+          update-only: true
+          message: |
+            ```
+            ${{ steps.validate_config.outputs.status_output }}
+            ```
```
Comment (hussainweb) [inline .github/workflows/ci.yml:182]: Can you explain this step better? It seems identical to the one before except that the condition is flipped.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -118,6 +121,74 @@ jobs:
...
+
+      - name: Fail if Config Export is not clean
+        if: ${{ !contains(steps.validate_config.outputs.status_output, 'No differences') }}
+        run: |
+          echo "Config export has differences"
+          exit 1
```
Comment (hussainweb) [inline .github/workflows/ci.yml:188]: We discussed that there are some edge cases where this could get reported. Since we know that, let's not abandon our tests. If GitHub Actions has a way to report warning (I vaguely remember something like that), let's do that. In any case, let's not stop executing the tests.

**Review by hussainweb — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 6 above)
Comment (hussainweb) [reply .github/workflows/ci.yml:144]: I know that `sort` needs to do additional work in a pipe but it's okay. In our case, it is not likely to get big. I'd rather remove extra lines here. That said, I don't feel strongly about this. It is fine the way it is right now as well.

Comment (zeshanziya) [reply .github/workflows/ci.yml:144]: @hussainweb I am removing this. I noticed `config:status` already produces sorted result. I confirmed checking drush code and there is a ksort method used. 

https://github.com/drush-ops/drush/blob/151b70439cdacda5b95546ca853f271d0fe981cf/src/Commands/config/ConfigCommands.php#L295

File: .github/workflows/ci.yml (hunk 7 above)
Comment (zeshanziya) [reply .github/workflows/ci.yml:182]: This updates the original comment that was added when there was a config difference. When we push missing configs again, on success, it will override the original comment. The `update-only: true` option ensures that the comment is updated but not added. For instance, on the first push only, if there are no config errors, it will not add any comment.

File: .github/workflows/ci.yml (hunk 8 above)
Comment (zeshanziya) [reply .github/workflows/ci.yml:188]: Ok sure, will implement this

File: .github/workflows/ci.yml (hunk 7 above)
Comment (hussainweb) [reply .github/workflows/ci.yml:182]: I see. What happens when there is another push to the PR which introduces a config change but is not exported?

Comment (zeshanziya) [reply .github/workflows/ci.yml:182]: It will update the comment with updated missing config list.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (zeshanziya) [reply .github/workflows/ci.yml]: Noted

File: .github/workflows/ci.yml (hunk 7 above)
Comment (zeshanziya) [reply .github/workflows/ci.yml]: @hussainweb I have merged both the steps into one to avoid confusion.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -118,6 +121,59 @@ jobs:
...
+            echo "EOF" >> $GITHUB_OUTPUT
+          else
+            echo "Config has no differences"
+            echo "status_output=No differences between DB and sync directory" >> $GITHUB_OUTPUT
+          fi
+
+      - name: Add/Update PR Comment
+        uses: mshick/add-pr-comment@v2
+        with:
+          update-only: ${{ contains(steps.validate_config.outputs.status_output, 'No differences') }}
```
Comment (hussainweb) [inline .github/workflows/ci.yml:162]: Let's explain this line with a comment.

**Review by hussainweb — CHANGES_REQUESTED** [review]
I need a small change but an important one.

**Review by zeshanziya — COMMENTED** [review]
Comment added explaining `update-only` flag

File: .github/workflows/ci.yml (hunk 9 above)
Comment (zeshanziya) [reply .github/workflows/ci.yml:164]: Changes are done
